### PR TITLE
State in docs that title/component/id must be statically defined

### PR DESCRIPTION
**Pull request by Sidnioulz**
Issue: if one uses a function to generate their stories' default export, the story file is skipped in SB 7, as there is no way to statically generate a title for it.

## What I did

I added a note in the documentation that helps SB 7 users make sense of this new behaviour. Having it indexed somewhere in the doc may help users find a self-service explanation through search engines when they run into the same issue as I did.

Besides that, an explicit migration note would help current users discover this new limitation without requiring live community support.

## How to test

- [ ] ~Is this testable with Jest or Chromatic screenshots?~
- [ ] ~Does this need a new example in the kitchen sink apps?~
- [ ] ~Does this need an update to the documentation?~

### CODE REVIEW

**Review by shilman — APPROVED** [review]
@Sidnioulz Thank you for the PR! The migration note in question is already here:

https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#string-literal-titles

However, I think your note is better and it deserves to be prominently displayed in the 7.0 notes. I've updated your PR to add it to the TOC, and will let @jonniebigodes review before merging.

**Review by shilman — COMMENTED** [review]
I've moved the migration note to 7.0 and also added instructions on how to opt out of the new behavior. This deserves some review @tmeasday since it might be redundant with other migration information. WDYT?

File: MIGRATION.md
Comment (Sidnioulz) [inline MIGRATION.md:286]: * [ ] Remove accidental copy/paste

File: MIGRATION.md
Comment (Sidnioulz) [inline MIGRATION.md:294]: * [ ] Remove accidental copy/paste

**Comment by Sidnioulz** [comment]
> I've moved the migration note to 7.0 and also added instructions on how to opt out of the new behavior. This deserves some review @tmeasday since it might be redundant with other migration information. WDYT?

Thanks for the extra commits! At first, I thought I had accidentally copy/pasted other doc elements. Happy to have this extra bit, though I think I could reuse the code examples in the other doc portion you pointed out, which were shorter. I'll rework the PR tomorrow.

**Comment by tmeasday** [comment]
I think that's great. The only thing I would add to it is some mention of `autotitle` which is designed to address (some of) the uses cases that dynamic title were used for.

**Comment by jonniebigodes** [comment]
Agree with @tmeasday on his comment. @Sidnioulz If you're ok with applying the suggestion as to add some clarification, I'd appreciate it.

**Comment by Sidnioulz** [comment]
Thanks for your patience, it's been a busy month IRL and I haven't had time to finish this earlier. I've added a link to the auto-title feature as requested!

**Comment by AasmundEndresen** [comment]
> > I've moved the migration note to 7.0 and also added instructions on how to opt out of the new behavior. This deserves some review @tmeasday since it might be redundant with other migration information. WDYT?
> 
> Thanks for the extra commits! At first, I thought I had accidentally copy/pasted other doc elements. Happy to have this extra bit, though I think I could reuse the code examples in the other doc portion you pointed out, which were shorter. I'll rework the PR tomorrow.

Sorry for commenting on an old issue here - but I thought it better then opening a new. This change is breaking our entire storybook as we are running a large and complex design system that relies on generating titles dynamically compile time, also to remove redundancies with out storySort function.

Its preventing us from being able to upgrade to Storybook 7 as the refactoring job to convert all our stories would be too large.

The autotitle feature also doesnt deliver enough flexibility to meet the usecase of dynamic titling as we're specifically using dynamic titles because we DON'T want the stories to be bound to the project structure or vice versa. Is it possible to open this conversation back up to look for solutions to how we can define a seperate visual structure for our stories and also be able to build our storySort function based on this structure?  I had previously made some utilities/tools that we've been using for this that relies on being able to use dynamic titles and id's.

Currently we're stuck between a rock and a hard place because Storybook 6.5 which we are currently using doesn't work on later versions of node then 16.9.0 so I have to switch back and forth node versions whenever we have to build our static storybook for production or run storybook, which is NOT ideal - and we can't upgrade to Storybook 7 because this breaking change you introduced by removing functionality. (I also didnt encounter the issue until recently because the automigration tool wasn't working for us until you released the 7.1 alpha, so we havent been able to get this far in the migration for half a year)

**Comment by tmeasday** [comment]
@AasmundEndresen I'll note that you can use `features: { storyStoreV7: false }` to likely work around this issue in SB7. However:

 - That isn't recommended for perf reasons
 - It will be deprecated soon and removed in SB8 so it'll be just buying time here.

As for you request, can you tell us a little more about what you are trying to achieve? Statically defined ids etc is very important but there may be ways to gain a bit more flexibility using e.g the [indexer API](https://github.com/storybookjs/storybook/discussions/23176#discussioncomment-6272118) or otherwise.

**Comment by Sidnioulz** [comment]
@tmeasday thanks for the indexer API RFC, I'll comment on it.

For me, specifically, I don't mind statically defined IDs at all, but story titles are the mechanism through which stories are categorised and sorted in the sidebar. So, I don't exactly need dynamic story titles, but a mechanism through which I can categorise / hierarchise stories and docs files. Having a separate, dynamic sidebar title could be a way to let users achieve that, whilst keeping the generation of stories.json fully static and pre-runtime.

**Comment by tmeasday** [comment]
@Sidnioulz if the issue is sorting, then the index needs to know about it, so changing the titles at "runtime" isn't really going to help.

If you aren't so much concerned about the titles or ids, why is `storySort` insufficient for your problem?

**Comment by Sidnioulz** [comment]
In v6, we could use a helper to get our users to categorise their components locally. We were able to ensure no typos were made on categories with TS. We then were able to sort categories with `storySort`.

```js
// In the CSF file

export default genMeta({
  category: Categories.FOO,
  component: 'Button',
})
// output: { title: 'Components/Foo/Button' }
```

I am not sure yet what a good v7 equivalent would be. I don't want to store each component category in my `main.js`, so I'd likely have to map stories to their story file and retrieve the default export of that file in `storySort`. Maybe 7.1's tags will offer a better solution, or maybe I can make do by injecting `category` into a custom parameter.

My goals are:
* Locally define, in the CSF, where the component will be categorised
* Enforce a strict list of categories, to avoid manually fixing typos by contributors
* Globally control how component categories are sorted

I don't think any of this is impossible in v7 but it doesn't feel as convenient.

**Comment by AasmundEndresen** [comment]
> @AasmundEndresen I'll note that you can use `features: { storyStoreV7: false }` to likely work around this issue in SB7. However:
> 
> * That isn't recommended for perf reasons
> * It will be deprecated soon and removed in SB8 so it'll be just buying time here.
> 
> As for you request, can you tell us a little more about what you are trying to achieve? Statically defined ids etc is very important but there may be ways to gain a bit more flexibility using e.g the [indexer API](https://github.com/storybookjs/storybook/discussions/23176#discussioncomment-6272118) or otherwise.

I've seen that solution in the docs, but I don't really consider it a solution as it'll be deprecated in SB8 and thus doesn't really solve the larger problem and just pushes back our problems to a later date.

I'll try to briefly explain our usecase - but since we run a large and complex design system it maybe would be more suitable to have a knowledge sharing session between our teams at some point to better explain the direction we've been trying to take our Storybook in.

The primary issue is maintenance - because we are as mentioned running a large design system for large organization, it involves alot of both documentation pages and components.  As @Sidnioulz said the autotitle and id functions just arent sufficient for creating a sensible information hierarchy at scale, so we've had to resolve our problems previously by creating a few utilities to our Storybook to help organize, index and maintain information.  Statically typing out the paths for each story creates a large risk of mistyping when there are alot of collaborators on the project, and our core team wont have the capacity to follow up and make sure every Pull Request we get have got this correctly.  We could include some steps in our pipeline to enforce this, but that would be silly as it would only add friction towards our collaborators. Bottom line is that we NEED to be able to use variables to create titles and id's to ensure consistency.

Furthermore we want to in a similar fashion be able to easily define and maintain mock data for our stories, and share configurations between stories - this is now also being hindered because the exported story object no longer can contain any form of variable defined elsewhere in the code. It creates maintenance hell - so its not really an option for us.

The performance have also never really been an issue for us, as we compile and statically build our Storybook anyways - so any variables or functions CAN be resolved before runtime, and in our local environments the performance have never been an issue either.

What I have done specifically for the titles and id's is create a central configuration for the hierarchy - I then import this configuration and use it in our storySort, and also have my utility that uses the same configuration in one of my custom made utilities to resolve any paths both for id's and titles based on that configuration. So sort of my own implementation of an autotitling functionality, but with more features that allws the user more control over how it works.

Similarly I have also created functionality to infer actions from the CustomEvents defined in our components (we use Stencil.js to generate web components) - and a function to automatically forward props, because since we're not working React or Vue we can't just use spread functionality, plus we also need to filter out undefined props.

Basically our requirement is to be able to share settings, configs, etc to keep our application DRY - and without being able to set the arguments of a Story based on variables and templates this becomes really difficult real quick.

The alternatives I'm considering right now is either look for something else that better suits our need, but I don't know what other packages that might be as Storybook is sort of the go-to for sandbox documentation environments.  Or we will have to write some kind of middleware and create an annoying inbetween step for running Storybook where we can configure our stories the way we like with variables and then compile them into a built 'stories' directory before running Storybook - this seems like the best option for now as it means we wont run the risk of future releases breaking us again, but it also seems at this point like we're halfway to the point where it would be better just building our own solution instead.
The third option I'm considering is exploring whether or not this is an area where it's possible to build an add-on, but I find the documentation for the Storybook API a bit lacking and don't fully understand if this is actually a possibility yet - though I feel like maybe this would be the ideal solution if its not possible to customize Storybook in future versions like this, atleast allowing addons to be built that can implement custom features in the sidebar or doing custom functionality with the args of a story.

**Comment by Sidnioulz** [comment]
> As @Sidnioulz said the autotitle and id functions just arent sufficient for creating a sensible information hierarchy at scale, so we've had to resolve our problems previously by creating a few utilities to our Storybook to help organize, index and maintain information. Statically typing out the paths for each story creates a large risk of mistyping when there are alot of collaborators on the project, and our core team wont have the capacity to follow up and make sure every Pull Request we get have got this correctly. We could include some steps in our pipeline to enforce this, but that would be silly as it would only add friction towards our collaborators. Bottom line is that we NEED to be able to use variables to create titles and id's to ensure consistency.

I echo that sentiment. At a large scale, things that can't be done through well-crafted helpers need to be done manually, which means a higher barrier to entry for contributors and more mental load for maintainers. This may sound silly for folks working in small systems, but it's a legitimate issue we're facing with contributions.

> The third option I'm considering is exploring whether or not this is an area where it's possible to build an add-on

This won't let you change titles, though you could comment on the linked Indexer API. Maybe the API could support editing the index rather than just adding to it.

--------

I'm hoping there's a way to add metadata to stories that I can then retrieve in `storiesSort`, which would allow me to slightly edit my helper to look like:

```js
export default {
  title: 'Statically defined',
  ...genMeta({
    category: Categories.FOO,
    otherMeta: 'bar'
  })
}
```

My understanding is SB 7.1 tags will be available and can help us restore such helpers.

**Comment by tmeasday** [comment]
@AasmundEndresen I'm not sure I understand this bit:

> Furthermore we want to in a similar fashion be able to easily define and maintain mock data for our stories, and share configurations between stories - this is now also being hindered because the exported story object no longer can contain any form of variable defined elsewhere in the code. It creates maintenance hell - so its not really an option for us.

Why can't you do this with CSF? It's only titles/names that aren't allowed to reference variables. All other keys of the CSF objects (barring tags, which is undocumented at this stage) can be as dynamic as you like.

**Comment by tmeasday** [comment]
> This won't let you change titles, though you could comment on the linked Indexer API. Maybe the API could support editing the index rather than just adding to it.

I mean you can write an indexer that consumes standard JS files and turns them into CSF compliant JS files (i.e. removes dynamism from the `title`/`name` fields) so I think it is technically possible to do what you want with an indexer. It's a bit convoluted though.

> I'm hoping there's a way to add metadata to stories that I can then retrieve in storiesSort, which would allow me to slightly edit my helper to look like:

Can't you use `parameters` for this?

> My understanding is SB 7.1 tags will be available and can help us restore such helpers

Tags are going to be static so might not be what you want here.

---

It seems to me that y'all really just want more powerful auto-titling. I'd need to hear a bit more about your `getMeta()` function but it seems to me that it would be easier and more maintainable to have something like:

```ts
// in main.js
stories: [
   { directory: './path/to/foo', prefix: 'Components/Foo' }
]
```

You can dynamically generate the `stories` field if you like 🤷

But maybe there's something lacking from autotitle that I'm not seeing. Maybe we can talk about that and see if we can come up with a way to make it fit the use case?

---

Another alternative if the problem is simply to "ensure no typos were made on categories with TS", to use https://github.com/storybookjs/storybook/pull/22791 or a linting rule?

**Comment by Sidnioulz** [comment]
> > I'm hoping there's a way to add metadata to stories that I can then retrieve in storiesSort, which would allow me to slightly edit my helper to look like:
> 
> Can't you use `parameters` for this?

No because it's not available to `storySort`, whereas tags are.

> > My understanding is SB 7.1 tags will be available and can help us restore such helpers
> 
> Tags are going to be static so might not be what you want here.

Bummer. We'd at least need to be able to import constants to feed into tags. If it's all done via static analysis on the AST again, then indeed it won't be helping my use case.

> It seems to me that y'all really just want more powerful auto-titling. I'd need to hear a bit more about your `getMeta()` function but it seems to me that it would be easier and more maintainable to have something like:
> 
> ```ts
> // in main.js
> stories: [
>    { directory: './path/to/foo', prefix: 'Components/Foo' }
> ]
> ```
> 
> You can dynamically generate the `stories` field if you like shrug

Nothing fancy in that function:

```js
export const genMeta = (options: MetaOptions): Concrete<Meta>['parameters'] => {
  const {
    category,
    description,
    name,
    ...rest
  } = options

  // We enforce a strict match with predefined categories.
  const categoryTitle = CATEGORIES[category]
  if (!categoryTitle) {
    throw new Error(
      `Invalid category: '${category}'. Valid categories are: ${Object.keys(
        CATEGORIES,
      ).join('; ')}.`,
    )
  }

 // Lots of irrelevant processing goes here

  return {
    ...rest,
    id: name,
    // Some more irrelevant processing elsewhere
    parameters: genParameters({
      ...options,
      context: CONTEXT.COMPONENT,
      description,
    }),
    // For all our component doc, category then name.
    title: `${categoryTitle}/${name}`,
  }
}
```
Modifying `main` like you suggest won't get me where I want. My goal is not, as an individual, to sort stories. My goal is to ensure that a large amount of contributors use well-defined and enforced constants to sort the individual story files they own into a large Storybook instance.

I'm considering something close, though. If I can add metadata to story files, or to my components in another way, I could load story files in main, and generate the stories config based on the metadata I read; essentially reimplementing dynamic titles by loading entire CSFs. This is just as awful as it sounds performance-wise, and will require serious tooling hacking to get my loaders supported in `main`'s context. At least it would let me preserve the public API for setting titles, with the layer of indirection I prefer to have.

> But maybe there's something lacking from autotitle that I'm not seeing. Maybe we can talk about that and see if we can come up with a way to make it fit the use case?

Happy to do that! Beyond this one specific topic, I'm wary of static analysis as it decreases the range of options we have to get stories written at scale. So if I can expose you to the kind of issues we're facing, and why such seemingly trivial details matter to us, all the better :)
 
> Another alternative if the problem is simply to "ensure no typos were made on categories with TS", to use #22791 or a linting rule?

That's a fallback I'm still considering, but not one I'm happy with. I have to let contributors hardcode category names in titles, I must indeed maintain a custom ESLint plugin to enforce categories, which is costlier than a set of constants and a utility function. 

Having the extra layer of abstraction afforded by the constants (which is an enum) allows me to fix typos in category names and to localise them, which wouldn't be doable if I had them written directly into titles.

**Comment by tmeasday** [comment]
> No because it's not available to storySort, whereas tags are.

> Bummer. We'd at least need to be able to import constants to feed into tags. If it's all done via static analysis on the AST again, then indeed it won't be helping my use case.

Ahh, yes of course, this is the chicken and egg problem. We are sorting the index statically because we need it ahead of loading the CSF files at runtime (so we on-demand load them). So we need to use static analysis to do that. Loading the CSF files to generate the index isn't really feasible for a whole lot of reasons I won't go into now.

> ...which is costlier than a set of constants and a utility function.

Did you look at the PR I linked? I don't know the TS string stuff well but it seems like it is just a enum/constant set of prefixes.

---

> Happy to do that! Beyond this one specific topic, I'm wary of static analysis as it decreases the range of options we have to get stories written at scale. So if I can expose you to the kind of issues we're facing, and why such seemingly trivial details matter to us, all the better :)

Like I said above, we are kind of stuck with static analysis. But I still don't really understand the problem. It seems like from what you've said:

- You want to generate titles based on a pattern
- The user provides a prefix and a name, but you want to check the prefix is from an allow list.

To me, it seems like autotitle would help here if the prefixes aren't arbitrary (i.e. all stories in one folder have a common prefix), and if the story filenames match the "name" part.

Otherwise you could do something tricky like ask them to put the category name in the story file name, so it would be `Button.foo.stories.ts`, and then your `stories` field for the `foo/` prefix could match `*.foo.stories.ts`? (I am imagining a scenario where you have various different kinds of stories for the same component in different files and those are the prefixes?)

**Comment by Sidnioulz** [comment]
> Did you look at the PR I linked? I don't know the TS string stuff well but it seems like it is just a enum/constant set of prefixes.

Cough cough... I did now :smile: I have to stay it's smarter than I thought. I suppose we could use a shim to globally extend `Meta` and enforce a title structure without trusting contributors to write the `satisfies` statement themselves. I'll def give it a try, thanks.

> To me, it seems like autotitle would help here if the prefixes aren't arbitrary (i.e. all stories in one folder have a common prefix), and if the story filenames match the "name" part.

The categorisation of components, in our org, may be based on business logic (they belong so the same family), or ownership (which tribe/squad maintains the component).

In our case, we have a monorepo structure and we don't want to complicate component import paths / package names to reflect on the categorisation. It's more about story sorting than about titles for us. It's just that story files is where it makes most sense to have the component owner store their metadata; and TS support is needed to ensure coherence.

In the long term, we also want to be able to plug in different metadata to the titles (sort by owner squad / by business domain / by status / by name). It's clear to me that this is something we won't be able to do now, though.

**Comment by tmeasday** [comment]
BTW One thing that we have planned with tags is to allow typing them, which I wonder if would help with your use case? ie. making it a type error to use a tag that isn't defined in `preview.js`.

**Comment by Sidnioulz** [comment]
> BTW One thing that we have planned with tags is to allow typing them, which I wonder if would help with your use case? ie. making it a type error to use a tag that isn't defined in preview.js.

Yes, it would help too! As far as my employer's concerned, we'll eventually embed Storybook into a larger-scale website and use stories.json to generate our own navigation, so we can fix story categorisation that way, and typed tags will provide a solid mechanism for us to do that.

In the meantime, we will still need to inject categories in the title, I suspect, as storySort only sorts and cannot hierarchise stories into categories itself. I'll likely use what you showed in the ticket you linked to handle this.

Thanks for taking the time to discuss this with us :)

**Pull request closed without merging**
